File: src/engine/DebugMetadata.ts
export class DebugMetadata {
  public startLineNumber: number = 0;
  public endLineNumber: number = 0;
  public startCharacterNumber: number = 0;
  public endCharacterNumber: number = 0;
  public fileName: string | null = null;
  public sourceName: string | null = null;

  public Merge(dm: DebugMetadata) {
    let newDebugMetadata = new DebugMetadata();

    newDebugMetadata.fileName = this.fileName;
    newDebugMetadata.sourceName = this.sourceName;

    if (this.startLineNumber < dm.startLineNumber) {
      newDebugMetadata.startLineNumber = this.startLineNumber;
      newDebugMetadata.startCharacterNumber = this.startCharacterNumber;
    } else if (this.startLineNumber > dm.startLineNumber) {
      newDebugMetadata.startLineNumber = dm.startLineNumber;
      newDebugMetadata.startCharacterNumber = dm.startCharacterNumber;
    } else {
      newDebugMetadata.startLineNumber = this.startLineNumber;
      newDebugMetadata.startCharacterNumber = Math.min(
        this.startCharacterNumber,
        dm.startCharacterNumber
      );
    }

    if (this.endLineNumber > dm.endLineNumber) {
      newDebugMetadata.endLineNumber = this.endLineNumber;
      newDebugMetadata.endCharacterNumber = this.endCharacterNumber;
    } else if (this.endLineNumber < dm.endLineNumber) {
      newDebugMetadata.endLineNumber = dm.endLineNumber;
      newDebugMetadata.endCharacterNumber = dm.endCharacterNumber;
    } else {
      newDebugMetadata.endLineNumber = this.endLineNumber;
      newDebugMetadata.endCharacterNumber = Math.max(
        this.endCharacterNumber,
        dm.endCharacterNumber
      );
    }

    return newDebugMetadata;
  }

  public toString() {
    if (this.fileName !== null) {
      return `line ${this.startLineNumber} of ${this.fileName}"`;
    } else {
      return "line " + this.startLineNumber;
    }
  }
}

File: ink_dart/lib/src/engine/debug_metadata.dart
/// Debug metadata for tracking source file locations.
///
/// Used for error reporting and debugging to map runtime content
/// back to original .ink source files.
class DebugMetadata {
  int startLineNumber = 0;
  int endLineNumber = 0;
  int startCharacterNumber = 0;
  int endCharacterNumber = 0;
  String? fileName;
  String? sourceName;

  DebugMetadata();

  /// Merge two metadata ranges into one that covers both.
  DebugMetadata merge(DebugMetadata other) {
    final newMetadata = DebugMetadata();

    // Use the earlier start
    if (startLineNumber < other.startLineNumber) {
      newMetadata.startLineNumber = startLineNumber;
      newMetadata.startCharacterNumber = startCharacterNumber;
    } else if (startLineNumber > other.startLineNumber) {
      newMetadata.startLineNumber = other.startLineNumber;
      newMetadata.startCharacterNumber = other.startCharacterNumber;
    } else {
      newMetadata.startLineNumber = startLineNumber;
      newMetadata.startCharacterNumber = startCharacterNumber < other.startCharacterNumber
          ? startCharacterNumber
          : other.startCharacterNumber;
    }

    // Use the later end
    if (endLineNumber > other.endLineNumber) {
      newMetadata.endLineNumber = endLineNumber;
      newMetadata.endCharacterNumber = endCharacterNumber;
    } else if (endLineNumber < other.endLineNumber) {
      newMetadata.endLineNumber = other.endLineNumber;
      newMetadata.endCharacterNumber = other.endCharacterNumber;
    } else {
      newMetadata.endLineNumber = endLineNumber;
      newMetadata.endCharacterNumber = endCharacterNumber > other.endCharacterNumber
          ? endCharacterNumber
          : other.endCharacterNumber;
    }

    newMetadata.fileName = fileName ?? other.fileName;
    newMetadata.sourceName = sourceName ?? other.sourceName;

    return newMetadata;
  }

  @override
  String toString() {
    if (fileName != null) {
      return 'line $startLineNumber of $fileName';
    }
    return 'line $startLineNumber';
  }
}
